modules/datadict/jsx/dataDictIndex.js

```diff
@@ -38,23 +40,94 @@ class DataDictIndex extends Component {
    * Called by React when the component has been rendered on the page.
    */
   componentDidMount() {
-    this.fetchData()
-      .then( () => this.setState({isLoaded: true}));
+      // Load the field options. This comes from a separate request than the
+      // table data stream. Once the fieldOptions are loaded, we set isLoaded
+      // to true so that the page is displayed with the data that's been
+      // retrieved.
+      fetch(this.props.fieldsURL, {credentials: 'same-origin'})
+          .then((resp) => resp.json())
+          .then((data) => this.setState({fieldOptions: data, isLoaded: true}))
+        .catch((error) => {
+            this.setState({error: true});
+            console.error(error);
+        });
+    this.fetchData();
   }
 
   /**
    * Retrive data from the provided URL and save it in state
    *
    * @return {object}
    */
-  fetchData() {
-    return fetch(this.props.dataURL, {credentials: 'same-origin'})
-        .then((resp) => resp.json())
-        .then((data) => this.setState({data}))
-        .catch((error) => {
-            this.setState({error: true});
-            console.error(error);
-        });
+    fetchData() {
+        const processLines = async (data) => {
+            const utf8Decoder = new TextDecoder('utf-8');
+            let row = [];
+            let colStart = -1;
+            let rowStart = 0;
+            for (let i = 0; i < data.length; i++) {
+                switch (data[i]) {
+                    case 0x1e: // end of column
+                        const rowdata = data.slice(colStart+1, i);
+                        const encoded = utf8Decoder.decode(rowdata);
+                        colStart = i;
+                        row.push(encoded);
+                        continue;
+                    case 0x1f: // end of row
+                        const rowdata2 = data.slice(colStart+1, i);
+                        const encoded2 = utf8Decoder.decode(rowdata2);
+                        row.push(encoded2);
+                        this.state.data.push(row);
```

Comment:
are you manipulating state directly here?

Reply:
I'll have to look into it.. I think this was intentional because I don't care if it re-renders after each row (it's probably preferable not to for performance reasons, as long as it's re-rendering at the end of each chunk returned by the reader, which it does below.)

Reply:
Would it be much work to alter this function so that it accepts a callback function to update state? that way we could extract this function from the component and it could be imported for use by other modules.

Reply:
this is done in the refactored version